### PR TITLE
example: Train Instruct pix2 pix with lora implementation

### PR DESCRIPTION
**Pull request by Aiden-Frost**
# What does this PR do?
Train the instructPix2Pix model on lower GPU resources using Lora.

## Who can review?
- Training examples: @sayakpaul and @patrickvonplaten

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your contributions. For starters, could you add it to the `research_projects` folder? If there's sufficient interest, we can graduate it to become an official script. 

Also, curious to see some results.

**Comment by Aiden-Frost** [comment]
Hello Sayak

I will move the train script to research_projects
Here are some simple statistics for training
dataset used: https://huggingface.co/datasets/instruction-tuning-sd/cartoonization



results

**Comment by sayakpaul** [comment]
I asked for results, actually.

**Comment by Aiden-Frost** [comment]
Currently, I have only fine-tuned the instructpix2pix model for around 100 epochs
By results, do you mean the edited image differences between the base instruct pix2pix model and the model fine-tuned with the cartoonization dataset?
Or is it something else?

**Comment by sayakpaul** [comment]
I mean can you show some results (input image, edit instruction, edited image) with the model finetuned with the technique you incorporated here?

**Comment by Aiden-Frost** [comment]
sure
Here are few examples


github_results drawio (1)

**Comment by sayakpaul** [comment]
Ah, nice! Thanks for sharing. 

Let's get code styling issues fixed. Then we can merge :)

**Review by patrickvonplaten — APPROVED** [review]

**Comment by sayakpaul** [comment]
Thanks for your contributions! Good job.

**Comment by DoctorTar** [comment]
sorry to bother but i met channel error in this script, the original train instruct pix2pix file add additional 4 channels in the unet to receive the original image embeddings. _https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix.py  line 453_
I tried to train pix2pix lora but failed because without this step it will throw error when tried to predict noise using unet.
_RuntimeError: Given groups=1, weight of size [320, 4, 3, 3], expected input[2, 8, 64, 64] to have 4 channels, but got 8 channels instead_
i guess we can not simply add the channel code back becase that would require training the whole unet and is not a lora training.
this channel issue happens since this script didnt add the channel part code, and when using unet to predict niose we simple cat the tensor together.
_concatenated_noisy_latents = torch.cat([noisy_latents, original_image_embeds], dim=1)_
Surely u tried ur scripts and it works, so i wonder why i failed for me both theoretically and realistically.

FYI i only change silghtly with the original code to successfully load my local data csv file and nothing else, i hardly think this would change anything.

**Comment by sayakpaul** [comment]
Cc: @Aiden-Frost

**Comment by Aiden-Frost** [comment]
Hello @DoctorTar, I just wanted to confirm: are you using the original instruct-pix2pix model `MODEL_ID="timbrooks/instruct-pix2pix"` ? This script was only meant to fine-tune that model since it has the in_channel modifications already done to it. Since the error mentioned that it was expected to have 4 channels, I am assuming you are using some other unet model (like stable-diffusion-v1-5). If this isn't the issue, please do ping me again. I would like to further debug this issue.


And regarding https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix.py line 453, I think I can implement that in this script such that anyone can train with any unet model. But its just odd that I am training the input layer and freezing the other layers. Like the new conv_in layer, it could learn new types of features that the layers below would have never seen before , thereby needing to train those layers as well (effectively the whole model). For this reason, the script I wrote only supported the fine-tuning of the instruct-pix2pix model.

If I do implement this feature, then during training, we train the unet.conv_in and the lora layers. I am just not sure if such training even makes sense (training just the input layer and the lora layers ). I would like to get some advice about this.

**Comment by DoctorTar** [comment]
> Hello @DoctorTar, I just wanted to confirm: are you using the original instruct-pix2pix model `MODEL_ID="timbrooks/instruct-pix2pix"` ? This script was only meant to fine-tune that model since it has the in_channel modifications already done to it. Since the error mentioned that it was expected to have 4 channels, I am assuming you are using some other unet model (like stable-diffusion-v1-5). If this isn't the issue, please do ping me again. I would like to further debug this issue.
> 
> And regarding https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix.py line 453, I think I can implement that in this script such that anyone can train with any unet model. But its just odd that I am training the input layer and freezing the other layers. Like the new conv_in layer, it could learn new types of features that the layers below would have never seen before , thereby needing to train those layers as well (effectively the whole model). For this reason, the script I wrote only supported the fine-tuning of the instruct-pix2pix model.
> 
> If I do implement this feature, then during training, we train the unet.conv_in and the lora layers. I am just not sure if such training even makes sense (training just the input layer and the lora layers ). I would like to get some advice about this.

OK, i did not use the specific model you montioned, i used  stable-diffusion-v1-5. just out of curiousity, is it the normal way (8 channels) when a model doing instruct-pix2pix task? which means the instruct-pix2pix model unet channel is always different from  other models ?

**Comment by Aiden-Frost** [comment]
> OK, i did not use the specific model you montioned, i used stable-diffusion-v1-5. just out of curiousity, is it the normal way (8 channels) when a model doing instruct-pix2pix task? which means the instruct-pix2pix model unet channel is always different from other models ?
> 
> In other words, instruct-pix2pix and img2img are completely 2 different things, while the former add addtional channels to embeds the original img and the latter add noise to the original img and denoise it , am i right? sorry if this is a basic/dumb question but i thought instruct-pix2pix and img2img were the same before.

So the model I have mentioned: `MODEL_ID="timbrooks/instruct-pix2pix"` is the model released by the [original paper](https://www.timothybrooks.com/instruct-pix2pix). Instructpix2pixis trained by using stable diffusion as its base model, along with the modifications done to the unet channel to support the image conditioning. You can get more detailed information from their [paper](https://arxiv.org/pdf/2211.09800.pdf) (section 3.2).

And regarding img2img and Instructpix2pix, correct me if I am wrong, but I do beleive the end result is similar (that is given an input image edit the image based on the text prompt); I just think their way of training and inference is slightly different. Even I am not 100% sure about it.
